Addition of bsp integration pages (#1551)
advancedtelematic/aktualizr: docs/ota-client-guide/modules/ROOT/pages/add-meta-updater-to-vendors-sdk.adoc

= Add meta-updater features to the vendor's SDK
ifdef::env-github[]

[NOTE]
====
We recommend that you link:https://docs.ota.here.com/ota-client/latest/{docname}.html[view this article in our documentation portal]. Not all of our articles render correctly in GitHub.
====
endif::[]

To avoid conflicts and bug classes, we require the use of `usrmerge`, even on systems that do not use systemd. In the https://www.yoctoproject.org/docs/2.7/mega-manual/mega-manual.html[Yocto] ecosystem, it is a general best practice to write `do_install()` functions that install to the variable `\{bindir}`, but some recipes still hard-code the location of `/bin`.

If a board vendor’s BSP has recipes like this, they will need to be patched so that the files install to the correct locations. In older Yocto versions, the only way to do this was to create a new BSP-specific layer with the patches, and only include that BSP-specific layer when the corresponding BSP was in use. (Otherwise, the recipes would fail to parse.)

> Currently, the https://patchwork.openembedded.org/patch/148271/[dynamic-layers] feature allows us to conditionally patch recipes from BSP layers if and only if they are present.

Reviewer: "Currently" is a weird word choice; it implies that the dynamic-layers feature might go away in later versions. In fact, it's the opposite--older Yocto versions didn't have dynamic layers, but newer ones do.

An example of a patch can be seen https://git.yoctoproject.org/cgit/cgit.cgi/meta-virtualization/tree/recipes-containers/cgroup-lite/cgroup-lite_1.15.bb[here], and this can be included in the meta-updater as the file `dynamic-layers/virtualization-layer/{asterisk}/{asterisk}/{asterisk}.bbappend` containing the following:

[source,bash]
----
do_install() {
	install -d ${D}/${bindir}
	install -m 0755 ${S}/scripts/cgroups-mount ${D}/${bindir}
	install -m 0755 ${S}/scripts/cgroups-umount ${D}/${bindir}
	install -d ${D}${sysconfdir}/init.d
	install -m 0755 ${WORKDIR}/cgroups-init ${D}${sysconfdir}/init.d/cgroups-init
	install -d ${D}${systemd_unitdir}/system
	ln -sf /dev/null ${D}${systemd_unitdir}/system/cgroups-init.service
}
----

> NOTE: You might need to modify a few installments in the `.bbappend` file, however the file needs to include the complete `do_install()` function.

Reviewer: Huh?

Author: another typo. "install statements" and not "installments".

Reviewer: Okay, now I think I see what this is about, and I think there's some domain knowledge you're missing here. A `.bbappend` file is like an amendment to a recipe: you can just modify some marginal parts, or add/change variables, or whatever, without having to fork the recipe or write your own version. Like, take a look at [this bbappend file](https://github.com/advancedtelematic/meta-updater/blob/master/recipes-connectivity/connman/connman_%25.bbappend) in meta-updater: it just adds one line of configuration, and leaves the rest of the recipe alone.

So let's say the `do_install()` function in the original recipe looks like this:

```bash
do_install() {
	install -d ${D}/bin
	install -m 0755 ${S}/scripts/cgroups-mount ${D}/bin
	install -m 0755 ${S}/scripts/cgroups-umount ${D}/bin
	install -d ${D}${sysconfdir}/init.d
	install -m 0755 ${WORKDIR}/cgroups-init ${D}${sysconfdir}/init.d/cgroups-init
	install -d ${D}${systemd_unitdir}/system
	ln -sf /dev/null ${D}${systemd_unitdir}/system/cgroups-init.service
}
```

You might think that, since you only need to fix the install statements that use /bin, your bbappend could do something like this:

```bash
do_install() {
	install -d ${D}/${bindir}
	install -m 0755 ${S}/scripts/cgroups-mount ${D}/${bindir}
	install -m 0755 ${S}/scripts/cgroups-umount ${D}/${bindir}
}
```
This won't work, though, because when you have a function in your bbappend it **replaces** the function from the parent file completely.

The reason for this NOTE to be here is to remind developers that if you want to modify anything inside the `do_install()` function, you have to replace the whole function, even if it's just one line in the function that needs to be modified. I'm not sure that's quite getting across.

Author: I think I get it now. I replaced this line with the line from the pdf as I think that probably explains it better.

== Add OTA Connect/meta-updater FS types

Meta-updater provides several FS types for secure OTA delivery which need to be added to the FS types for the image class being built.

For example, in the `fsl-image-networking-full.bb` image file, the line below specifies the image type to be built.
----
IMAGE_FSTYPES_qoriq = "tar.gz ext2.gz.u-boot ext2.gz"
----

This image builds `tar.gz`, `ext2.gz.u-boot`, and `ext2.gz` by default.

> We can also decide to add new FS types such as `ostreepush`, `garagesign`, `garagecheck`, `ota-ext4`, `ostree.tar.bz2`, `ota.tar.xz` and `wic` with the line below.

Reviewer: Don't use "can" here; these aren't optional.

= Add meta-updater features to the vendor's SDK
ifdef::env-github[]

[NOTE]
====
We recommend that you link:https://docs.ota.here.com/ota-client/latest/{docname}.html[view this article in our documentation portal]. Not all of our articles render correctly in GitHub.
====
endif::[]

> To avoid conflicts and bug classes, we require the use of `usrmerge`, even on systems that do not use systemd. In the https://www.yoctoproject.org/docs/2.7/mega-manual/mega-manual.html[Yocto] ecosystem, it is a general best practice to write `do_install()` functions that install to the variable `\{bindir}`, but some recipes still hard-code the location of `/bin`.

Reviewer 2: ```suggestion
To avoid conflicts and certain classes of bugs, we require the use of `usrmerge`, even on systems that do not use systemd. In the https://www.yoctoproject.org/docs/2.7/mega-manual/mega-manual.html[Yocto] ecosystem, it is a general best practice to write `do_install()` functions that install to the variable `\{bindir}`, but some recipes still hard-code the location of `/bin`.

```

[source,bash]
----
do_install() {
	install -d ${D}/${bindir}
	install -m 0755 ${S}/scripts/cgroups-mount ${D}/${bindir}
	install -m 0755 ${S}/scripts/cgroups-umount ${D}/${bindir}
	install -d ${D}${sysconfdir}/init.d
	install -m 0755 ${WORKDIR}/cgroups-init ${D}${sysconfdir}/init.d/cgroups-init
	install -d ${D}${systemd_unitdir}/system
	ln -sf /dev/null ${D}${systemd_unitdir}/system/cgroups-init.service
}
----

NOTE: Even though only some of the install statements needed to be modified, the `bbappend` file still needs to include the complete `do_install()` function.

== Add OTA Connect/meta-updater FS types

Meta-updater provides several FS types for secure OTA delivery which need to be added to the FS types for the image class being built.

For the LS1043ARDB, we targeted the `fsl-image-networking-full.bb` image file, the line below specifies the image type to be built.
[source,bash]
----
IMAGE_FSTYPES_qoriq = "tar.gz ext2.gz.u-boot ext2.gz"
----

This image builds `tar.gz`, `ext2.gz.u-boot`, and `ext2.gz` by default.

> You should add the following FS types `ostreepush`, `garagesign`, `garagecheck`, `ota-ext4`, `ostree.tar.bz2`, `ota.tar.xz` and `wic` with the line below.

Reviewer 2: Do we actually need all of these? @eu-smirnov can you confirm this?

For the LS1043ARDB, this is how we implemented the script. Note that, because it was very simple, we elected to simply embed the text of the file in the `deploy_append` of the qoriq image; this could also be done via including files, but it would be a bit more roundabout that way. This `bbappend` just cats the 4 lines of the script into a file in bitbake’s deployment directory, and then places it in the image.

[source,bash]
----
DEPENDS += "u-boot-mkimage-native"

do_deploy_append_qoriq() {
	cat > ${DEPLOYDIR}/ls1043ardb_boot.txt << EOF
load mmc 0:2 \${load_addr} /boot/loader/uEnv.txt <1>
env import -t \${fileaddr} \${filesize}
load mmc 0:2 \${load_addr} /boot\${kernel_image}
bootm \${load_addr}
EOF

	mkimage -A arm64 -O linux -T script -d ${DEPLOYDIR}/ls1043ardb_boot.txt ${DEPLOYDIR}/ls1043ardb_boot.scr <2>
}
----

<1> This is the crucial line pointing the bootloader to the OSTree-managed script. The OSTree-managed script sets the values of `$\{fileaddr}`, `$\{filesize}`, and `$\{kernel_image}`.

> <2> Note that `ls1043ardb_boot.scr` is the name of the file specified to be included as `IMAGE_BOOT_FILES` in the boardspecific bbclass described in the xref:add-board-class.adoc[first step].

Reviewer 2: ```suggestion
<2> Note that `ls1043ardb_boot.scr` is the name of the file specified to be included as `IMAGE_BOOT_FILES` in the board-specific bbclass described in the xref:add-board-class.adoc[first step].

```

Several recipes needed to be patched in our example. One example patch can be seen here of the  https://git.yoctoproject.org/cgit/cgit.cgi/meta-virtualization/tree/recipes-containers/cgroup-lite/cgroup-lite_1.15.bb[cgroup-lite-recipe], and this can be included in the meta-updater as the file `dynamic-layers/virtualization-layer/recipes-containers/cgroup-lite/cgroup-lite_%.bbappend`:

[source,bash]
----
do_install() {
	install -d ${D}/${bindir}
	install -m 0755 ${S}/scripts/cgroups-mount ${D}/${bindir}
	install -m 0755 ${S}/scripts/cgroups-umount ${D}/${bindir}
	install -d ${D}${sysconfdir}/init.d
	install -m 0755 ${WORKDIR}/cgroups-init ${D}${sysconfdir}/init.d/cgroups-init
	install -d ${D}${systemd_unitdir}/system
	ln -sf /dev/null ${D}${systemd_unitdir}/system/cgroups-init.service
}
----

NOTE: Even though only some of the install statements needed to be modified, the `bbappend` file still needs to include the complete `do_install()` function.

== Add OTA Connect/meta-updater FS types

Meta-updater provides several FS types for secure OTA delivery which need to be added to the FS types for the image class being built.

> For the LS1043ARDB, we targeted the `fsl-image-networking-full.bb` image file, the line below specifies the image type to be built.

Reviewer: ```suggestion
For the LS1043ARDB, we targeted the `fsl-image-networking-full.bb` image file, the line below specifies the image types to be built.
```

Just a missing plural.
